File: Patterns/Command Pattern/CSharp/Command/Light/LightOnCommand.cs
namespace Command.Light
{
    public class LightOnCommand : ICommand
    {
        private Light light;
        private int preview;

        public LightOnCommand(Light light)
        {
            preview = Light.OFF;
            this.light = light;
        }

        public void Execute()
        {
            preview = light.State;
            light.On();
        }

        public void Undo()
        {
            if (preview == Light.OFF)
            {
                light.Off();
            }
        }
    }
}

File: Patterns/Command Pattern/Cpp/Command/Light/LightOnCommand.cpp
#include "../ICommand.cpp"
#include "./Light.cpp"

class LightOnCommand : public ICommand {
private:
    Light* light;
    int preview;

public:
    LightOnCommand(Light* light) : light{light}, preview{light->OFF} {}

    void execute() override {
        preview = light->getState();
        light->on();
    }

    void undo() override {
        if (preview == light->OFF) light->off();
    }

    string getCommandName() override {
        return "LightOnCommand";
    }
};